src/client/Microsoft.Identity.Client/Region/RegionDiscovery.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Internal.Logger;
+using Microsoft.Identity.Client.PlatformsCommon.Shared;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.Region
+{
+    internal sealed class RegionDiscovery
+    {
+        private const string RegionName = "REGION_NAME";
+        private Uri ImdsUri;
+        private IDictionary<string, string> Headers;
+        private ICoreLogger Logger;
+        private static readonly object _lock = new object();
+        private static RegionDiscovery Instance = null;
+        private HttpManager HttpManager;
+
+        private RegionDiscovery()
+        {
+            ImdsUri = new Uri("http://169.254.169.254/metadata/instance/compute");
+            Headers = new Dictionary<string, string>();
+            Headers.Add("api-version", "2019-06-01");
+            Logger = new NullLogger();
+            HttpManager = new HttpManager(new SimpleHttpClientFactory());
+        }
+
+        // For Unit tests only.
+        public void setHttpManager(HttpManager httpManager)
+        {
+            HttpManager = httpManager;
+        }
+
+        internal static RegionDiscovery GetInstance
```

Comment:
Please avoid singletons if possible, they are hard to test. In this case I would put all this logic inside the DiscoveryManager as a provider.

Reply:
changed